### DISCUSSION TITLE
ComparisonIndexer is returning a NPE

### DISCUSSION THREAD

**Discussion by mikepictor** [TimefoldAI/timefold-solver#632] [General]
I am getting a NPE from the ComparisonIndexer, not always, but often enough. It happens here..
```
    public ElementAwareListEntry<T> put(IndexProperties indexProperties, T tuple) {
        Key_ indexKey = indexProperties.toKey(propertyIndex);
        // Avoids computeIfAbsent in order to not create lambdas on the hot path.
        Indexer<T> downstreamIndexer = comparisonMap.get(indexKey);
        if (downstreamIndexer == null) {
            downstreamIndexer = downstreamIndexerSupplier.get();
            comparisonMap.put(indexKey, downstreamIndexer);
        }
        return downstreamIndexer.put(indexProperties, tuple);
    }
```
In my case, propertyIndex is 2, indexKey is null, indexProperties has 2 elements in it (so the index of 2 points to nothing, thus the null, then on the `downstreamIndexer.put` it tries to place the tuple with a null key, leading to a NPE from the java Comparators when it compare null to the other keys in the map.

I don't know enough about how this is meant to work, or why the propertyIndex is 2 when there are only properties at index 0 and 1
![image](https://github.com/TimefoldAI/timefold-solver/assets/4552747/88f802af-447b-40a3-8b54-e5769fe03f8e)

I can't tell if this is a timefold bug, whether my planning entities are badly formed, or something else.

**Comment by mikepictor** [comment]
Or indeed, maybe it's my constraint that is causing an issue, though I don't feel they are overly complex constraints.

**Comment by triceo** [comment]
Hello. Thanks for bringing this to our attention.
Can you please provide a reproducer? Something I can run on my machine to see your problem?
Without that, I'm afraid I can't do much to help.
